File: Java-Core/src/main/java/presto/grammar/VariableInstance.java
package presto.grammar;

import presto.error.PrestoError;
import presto.error.SyntaxError;
import presto.expression.IExpression;
import presto.runtime.Context;
import presto.runtime.Variable;
import presto.type.DocumentType;
import presto.type.IType;
import presto.utils.CodeWriter;
import presto.value.IValue;

public class VariableInstance implements IAssignableInstance {
	
	Identifier name;
	
	public VariableInstance(Identifier name) {
		this.name = name;
	}
	
	public Identifier getName() {
		return name;
	}
	
	@Override
	public void toDialect(CodeWriter writer) {
		writer.append(name);
	}
	
	@Override
	public void checkAssignValue(Context context, IExpression expression) throws SyntaxError {
		IType type = expression.check(context);
		INamed actual = context.getRegisteredValue(INamed.class,name);
		if(actual==null)
			context.registerValue(new Variable(name, type));
		else {
			// need to check type compatibility
			IType actualType = actual.getType(context);
			type.checkAssignableTo(context,actualType);
		}
	}
	
	@Override
	public void checkAssignMember(Context context, Identifier memberName) throws SyntaxError {
		INamed actual = context.getRegisteredValue(INamed.class,name);
		if(actual==null) 
			throw new SyntaxError("Unknown variable:" + this.name);
		if(actual.getType(context)!=DocumentType.instance())
			throw new SyntaxError(this.name + " is not a document. Cannot assign member!");
	}
	
	@Override
	public void checkAssignElement(Context context) throws SyntaxError {
		// TODO Auto-generated method stub
		
	}
	
	@Override
	public void assign(Context context, IExpression expression) throws PrestoError {
 		IValue value = expression.interpret(context);
		if(context.getRegisteredValue(INamed.class,name)==null)
			context.registerValue(new Variable(name, value.getType())); 
		context.setValue(name, value);
	}
	
	@Override
	public IValue interpret(Context context) throws PrestoError {
		return context.getValue(name);
	}

}

File: Java-CodeServer/src/main/resources/js/prompto/grammar/VariableInstance.js
var Variable = require("../runtime/Variable").Variable;
var SyntaxError = require("../error/SyntaxError").SyntaxError;
var DocumentType = require("../type/DocumentType").DocumentType;

function VariableInstance(name) {
	this.name = name;
	return this;
}

VariableInstance.prototype.toDialect = function(writer, expression) {
    if(expression!=null) try {
        var type = expression.check(writer.context);
        var actual = writer.context.getRegisteredValue(name);
        if(actual==null)
            writer.context.registerValue(new Variable(name, type));
    } catch(e) {
        // TODO warning
    }
    writer.append(this.name);
};

VariableInstance.prototype.toString = function() {
    return this.name;
};

VariableInstance.prototype.checkAssignValue = function(context, expression) {
	var type = expression.check(context);
	var actual = context.getRegisteredValue(this.name);
	if(actual==null) {
		expression.check(context);
		context.registerValue(new Variable(this.name, type));
	} else {
		// need to check type compatibility
		type.checkAssignableTo(context,actual.type);
	}
};

VariableInstance.prototype.checkAssignMember = function(context, memberName) {
	var actual = context.getRegisteredValue(this.name);
	if(actual==null) {
		throw new SyntaxError("Unknown variable:" + this.name);
	}
};

/*

@Override
public void checkAssignElement(Context context) throws SyntaxError {
	// TODO Auto-generated method stub
	
}

*/
VariableInstance.prototype.assign = function(context, expression) {
	var value = expression.interpret(context);
	if(context.getRegisteredValue(this.name)==null) {
		context.registerValue(new Variable(this.name, value.type));
	}
	context.setValue(this.name, value);
};

VariableInstance.prototype.interpret = function(context) {
	return context.getValue(this.name);
};

exports.VariableInstance = VariableInstance;